[ADD] estate: create real estate module

clom-odoo opened the pull request:
Creation of a module for real estate.
There should be more description however it's work in progress

clom-odoo committed:
[ADD] estate: create real estate module

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/783.png)](https://mergebot.odoo.com/odoo/tutorials/pull/783)

clom-odoo committed:
[IMP] estate: add a model for estate properties

clom-odoo committed:
[IMP] estate: add some basic security

clom-odoo committed:
[IMP] estate: added some menus and updated fields and defaults value

clom-odoo committed:
[IMP] estate: customized the search bar with fields, filters and group_by

clom-odoo committed:
[IMP] estate: add relational fields about property Type, Tags and Offers

clom-odoo committed:
[IMP] estate: update some fields to be computed and add onchange basics settings

clom-odoo committed:
[IMP] estate: add some actions to cancel or accept offer for properties and set properties as sold/cancelled

clom-odoo committed:
[IMP] estate: add some sql and python constraints

clom-odoo committed:
[IMP] estate: add some color, ordering and stats buttons

clom-odoo committed:
[IMP] estate: add some protection on create/delete operation using inheritance and add a page in User to see Properties linked to salesperson

clom-odoo committed:
[ADD] estate_account: module to create invoice from properties when sold

clom-odoo committed:
[IMP] estate: add a not draggable Kanban view for estate_property

clom-odoo committed:
[REF] estate, estate_account: change names according to the coding guideline and clear some typos

This commit will refactor estate_property in estate.property as recommended. Some file name especially those who deals with inhertiance are changed accoridng to the guideline.

ushyme commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,20 @@
+    offer_ids = fields.One2many("estate.property.offer", "property_type_id")
+    offer_count = fields.Integer(compute="_count_offer")
+
+    _sql_constraints = [("unique_property_tag", "UNIQUE(name)", "Tag name must be unique !")]
```
```suggestion
    _sql_constraints = [("unique_property_type", "UNIQUE(name)", "Tag name must be unique !")]
```
 Constraint name `"unique_property_tag"` is misleading

ushyme commented on estate/models/estate_property_tag.py:
```diff
@@ -0,0 +1,12 @@
+from odoo import fields, models
+
+
+class PropertyTags(models.Model):
```
```suggestion
class PropertyTag(models.Model):
```
Model class names should be singular

ushyme commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,99 @@
+    @api.depends("offer_ids")
+    def _find_best_price(self):
+        for record in self:
+            record.best_price = max(record.offer_ids.mapped("price") + [0])
```
```suggestion
            record.best_price = max(record.offer_ids.mapped("price"), default=0)
```
`max()` lets you set a default value

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,67 @@
+    ]
+
+    @api.depends("validity")
+    def _compute_deadline(self):
+        for record in self:
+            if record.create_date:
+                now = record.create_date
+                record.date_deadline = now + relativedelta(days=record.validity)
```
This will crash if there's no `create_date`

ushyme commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,67 @@
+            if offer.create_date:
+                offer.validity = (offer.date_deadline - offer.create_date.date()).days
+
+    def action_confirm(self):
+        for offer in self:
+
+            for other_offer in offer.property_id.offer_ids:
+                if other_offer.status == "accepted":
+                    offer.status = "refused"
+
+            if offer.status == "refused":
+                raise UserError("Offer already refused")
+
+            offer.status = "accepted"
+            offer.property_id.buyer = offer.partner_id
+            offer.property_id.selling_price = offer.price
+            for other_offer in offer.property_id.mapped("offer_ids"):
+                if other_offer.id != offer.id:
+                    other_offer.status = "refused"
+        return True
```
```suggestion
    def action_confirm(self):
        for offer in self:
            if offer.status == 'refused':
                raise UserError("You cannot accept an offer that is already refused.")

            other_offers = offer.property_id.offer_ids.filtered(lambda o: o.id != offer.id and o.status != 'refused')
            other_offers.write({'status': 'refused'})

            offer.write({'status': 'accepted'})
            offer.property_id.write({
                'buyer': offer.partner_id.id,
                'selling_price': offer.price,
            })

        return True

```

ushyme commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,54 @@
+        <field name="name">estate.property.offer.list</field>
+        <field name="model">estate.property.offer</field>
+        <field name="arch" type="xml">
+            <list string="Channel">
```
"Channel" ? There are more of these in other files.

ushyme commented on estate/views/estate_property_tag_views.xml:
```diff
@@ -0,0 +1,44 @@
+        <field name="name">estate.property.tag.form</field>
+        <field name="model">estate.property.tag</field>
+        <field name="arch" type="xml">
+            <form string="Property">
```
```suggestion
            <form string="Property Tag">
```

ushyme commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,54 @@
+        <field name="name">estate.property.offer.form</field>
+        <field name="model">estate.property.offer</field>
+        <field name="arch" type="xml">
+            <form string="Property">
```
```suggestion
            <form string="Property Offer">
```

ushyme commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,25 @@
+        res = super().action_sold()
+        self.env['account.move'].create({
+            "move_type": 'out_invoice',
+            "partner_id": self.buyer.id,
```
What if the `self.buyer` doesn't exist ?

clom-odoo replied:
As buyer is a many2one field it gets a blank template by default. Therefore there is a field `buyer.id` with the value `False` which is interpreted as no buyer.

So when there is no buyer, this will create an invoice with no Customer

ushyme commented on estate/views/estate_property_type_views.xml:
```diff
@@ -0,0 +1,71 @@
+                    <button class="oe_stat_button"
+                        name="estate.estate_property_offer_action_filter"
+                        type="action" string="test">
+                        <!-- 
+                        <div class="o_field_widget o_stat_info">
+                            <span class="o_stat_value d-flex gap-1">
+                                <field name="offer_count" widget="statinfo" nolabel="1" class="oe_inline"/>
+                                <field name="name" class="oe_inline"/>
+                            </span>
+                        </div>
+                        -->
```
```suggestion
```
In general, we try to use comments to add context to the code. This doesn't seem to add much.

clom-odoo replied:
I forgot to uncomment it after debugging

ushyme commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,149 @@
+                <field string="Bedrooms" name="bedrooms"/>
+                <field string="Living Area (sqm)" name="living_area" filter_domain="[('living_area', '&gt;', self)]"/>
+                <field string="Facades" name="facades"/>
+                <filter string="Available" name="state_availability_filter" domain="['|',('state','=', 'new'), ('state','=','offer received')]"/>
```
Domains have a `in` operator

ushyme reviewed (changes requested):
Good work overall. One general remark: There are inconsistencies with the use of single/double quotes. We generally use different quote signs to distinguishes between technical strings (single quotes) and user-facing strings (double quotes).

clom-odoo committed:
[IMP] estate, estate_account: correct typo and refactor according to the review

Add a lambda fonction to the default field computing a date based on fields.Date.today method.

Add a line at the end of file for files where it was missing.

Adapted the create method of the EstatePropertyOffer to match the api.model_create_multi decorator.

Add a call to the self.env._ method at each UserError callto export them for translation.

Updated class name according to the guideline to match their model name in CamelCase format.

Updated 'offer received' and 'offer accepted' to snake_case formatting.

Updated misleading _sql_constraints name and error from estate_property_type.

Switch to default return value instead of a workaround for max method in estate/models/estate_property.py .

Updated action_confirm to:
1) right algorithm
2) better readability using write and filtered method

Corrected some string field in xml.

Uncommented the core of a stat_button that was only displaying 'test'.
